### PR TITLE
Added gray image support to `adjust_saturation` function

### LINKED ISSUE #4466: Support grayscale image-like tensors for adjust_contrast and adjust_saturation functions
### 🚀 The feature

The following functions in `functional_tensor.py` should be updated to work with single-channel image-like tensor, i.e. `(1, H, W)`.
1. `adjust_contrast`. Similar implementation in Pillow is:
```python
    enhancer = ImageEnhance.Contrast(img)
    img = enhancer.enhance(contrast_factor)
```
2. `adjust_saturation`. Similar implementation in Pillow is:
```python
    enhancer = ImageEnhance.Color(img)
    img = enhancer.enhance(saturation_factor)
```

### Motivation, pitch

When updating the documentation, I discovered some transform functions that work well on one channel PIL image (image with `mode='L'`), yet they only support 3-channel tensors as in #4457.

Therefore, I think it is necessary to re-check and implement those functions to provide a similar result as in transforming PIL images.

Also, after solving this issue, we can further enable users to use `AutoAugment` and similar classes to work with grayscale image tensors. Currently, grayscale pillow images can use these augmentations without errors, but not with single-channel image tensors, i.e. `(1,H,W)`

### Alternatives

_No response_

### Additional context

_No response_

cc @vfdev-5 @datumbox

### PR DESCRIPTION
**Pull request by VinhLoiIT**
This resolves #4466 

- [x] Update code for `adjust_contrast`
- [x] Update code for `adjust_saturation`
- [x] Update document for `adjust_contrast`
- [x] Update document for `adjust_saturation`

cc @vfdev-5 @datumbox

### CODE REVIEW

**Comment by VinhLoiIT** [comment]
I found that `ColorJitter` depends on this issue as well while reviewing the documentation, and currently, the tests just run on 3-channel images.
https://github.com/pytorch/vision/blob/92524362900fc70ea920c3fb5a9347fa9b4dfc5c/test/test_transforms_tensor.py#L70

**Comment by vfdev-5** [comment]
@VinhLoiIT I think we can do the same trick with channels there

**Comment by vfdev-5** [comment]
For example:
```python
@pytest.mark.parametrize('device', cpu_and_gpu())
@pytest.mark.parametrize('channels', [3, 1])
class TestColorJitter:

    @pytest.mark.parametrize('brightness', [0.1, 0.5, 1.0, 1.34, (0.3, 0.7), [0.4, 0.5]])
    def test_color_jitter_brightness(self, brightness, device, channels):
        tol = 1.0 + 1e-10
        meth_kwargs = {"brightness": brightness}
        _test_class_op(
            T.ColorJitter, meth_kwargs=meth_kwargs, test_exact_match=False, device=device,
            tol=tol, agg_method="max", channels=channels
        )
    # same of other tests
```

cc @NicolasHug is it OK to do that like that ?

**Comment by NicolasHug** [comment]
> cc @NicolasHug is it OK to do that like that with parametrize on TestColorJitter as for device ?

Sorry I'm missing some context. If you're asking whether we can add a parametrization for `channel` at the class level, then yes, as long as `channel` is an argument of all the methods :)

**Comment by vfdev-5** [comment]
> > cc @NicolasHug is it OK to do that like that with parametrize on TestColorJitter as for device ?
> 
> Sorry I'm missing some context. If you're asking whether we can add a parametrization for `channel` at the class level, then yes, as long as `channel` is an argument of all the methods :)

yes, I mean do we have any restrictions on what we parametrize and where. Device parameter is something very generic and applicable to everyone. Channel as parameter is rather specific to `TestColorJitter`. Even if we technically do as suggested in the example, we may have some thought on how to parametrize and where.

**Comment by NicolasHug** [comment]
> do we have any restrictions on what we parametrize and where

Not really, we should just consider parametrizations to be glorified for-loops. If it makes sense to run a test with different values, then parametrization is relevant.

In this specific case of `TestColorJitter`, since the tests are testing different sub-transforms, I think it makes sense to add the parametrization at the class level.

File: test/test_transforms_tensor.py [hunk 1]
```diff
@@ -109,109 +109,118 @@ def _test_op(func, method, device, fn_kwargs=None, meth_kwargs=None, test_exact_
...
+def test_random(func, method, device, channels, fn_kwargs, match_kwargs):
+    _test_op(func, method, device, channels, fn_kwargs, fn_kwargs, **match_kwargs)
 
 
 @pytest.mark.parametrize('device', cpu_and_gpu())
 class TestColorJitter:
 
     @pytest.mark.parametrize('brightness', [0.1, 0.5, 1.0, 1.34, (0.3, 0.7), [0.4, 0.5]])
-    def test_color_jitter_brightness(self, brightness, device):
+    @pytest.mark.parametrize('channels', [1, 3])
```
Comment (vfdev-5) [inline test/test_transforms_tensor.py:121]: Let's put channels to the class as done for `device`

Comment (VinhLoiIT) [reply test/test_transforms_tensor.py:121]: Oh, I see. It's definitely better.

File: test/test_transforms_tensor.py [hunk 2]
```diff
@@ -109,109 +109,118 @@ def _test_op(func, method, device, fn_kwargs=None, meth_kwargs=None, test_exact_
...
-            tol=12.1, agg_method="max"
+            tol=12.1, agg_method="max", channels=channels
         )
 
 
 @pytest.mark.parametrize('device', cpu_and_gpu())
 @pytest.mark.parametrize('m', ["constant", "edge", "reflect", "symmetric"])
 @pytest.mark.parametrize('mul', [1, -1])
-def test_pad(m, mul, device):
+@pytest.mark.parametrize('channels', [1, 3])
```
Comment (vfdev-5) [inline test/test_transforms_tensor.py:172]: I do not think that we need to add specific channel 1 test to geometric transforms like pad, crop etc

Comment (VinhLoiIT) [reply test/test_transforms_tensor.py:168]: Yeah, it's a little bit confusing to me at first because it's quite duplicated.

It's okay to remove and leave channels=3 as default, though.

File: test/test_transforms_tensor.py [hunk 3]
```diff
@@ -483,7 +486,7 @@ def test_to_grayscale(device, Klass, meth_kwargs):
     tol = 1.0 + 1e-10
     _test_class_op(
         Klass, meth_kwargs=meth_kwargs, test_exact_match=False, device=device,
-        tol=tol, agg_method="max"
+        tol=tol, agg_method="max", channels=3,
```
Comment (vfdev-5) [inline test/test_transforms_tensor.py:489]: I think we can revert this as channels 3 is the defaults value.

File: test/test_transforms_tensor.py [hunk 4]
```diff
@@ -194,7 +196,8 @@ def test_crop(device):
     # Test transforms.RandomCrop with size and padding as tuple
     meth_kwargs = {"size": (4, 5), "padding": (4, 4), "pad_if_needed": True, }
     _test_op(
-        F.crop, T.RandomCrop, device=device, fn_kwargs=fn_kwargs, meth_kwargs=meth_kwargs
+        F.crop, T.RandomCrop, device=device,
+        fn_kwargs=fn_kwargs, meth_kwargs=meth_kwargs
```
Comment (vfdev-5) [inline test/test_transforms_tensor.py:200]: Let's also revert this change.

**Review by vfdev-5 — COMMENTED** [review]
Thanks for the PR @VinhLoiIT 
I left few nit comments, please fix them.

**Comment by VinhLoiIT** [comment]
I don't know whether to update the documentation as well as channel testing of `AutoAug` and related classes in #4457 inside this PR or to create another PR for that?

File: torchvision/transforms/transforms.py [hunk 5]
```diff
@@ -1098,7 +1098,7 @@ def __repr__(self):
 class ColorJitter(torch.nn.Module):
     """Randomly change the brightness, contrast, saturation and hue of an image.
     If the image is torch Tensor, it is expected
-    to have [..., 3, H, W] shape, where ... means an arbitrary number of leading dimensions.
+    to have [..., 1 or 3, H, W] shape, where ... means an arbitrary number of leading dimensions.
```
Comment (vfdev-5) [inline torchvision/transforms/transforms.py:1101]: > If img is PIL Image, mode "1", "L", "I", "F" and modes with transparency (alpha channel) are not supported.

"L" mode is now supported for PIL images, right ?

File: torchvision/transforms/functional.py [hunk 6]
```diff
@@ -842,7 +842,7 @@ def adjust_hue(img: Tensor, hue_factor: float) -> Tensor:
 
     Args:
         img (PIL Image or Tensor): Image to be adjusted.
-            If img is torch Tensor, it is expected to be in [..., 3, H, W] format,
+            If img is torch Tensor, it is expected to be in [..., 1 or 3, H, W] format,
```
Comment (vfdev-5) [inline torchvision/transforms/functional.py:845]: > If img is PIL Image mode "1", "L", "I", "F" and modes with transparency (alpha channel) are not supported.

Same here, "L" is not now support

Comment (VinhLoiIT) [reply torchvision/transforms/functional.py:845]: On the pillow side, the `mode="L"` is supported before this PR as you can see in
https://github.com/pytorch/vision/blob/011eedfbf6a86d22da443141139519493e33eec5/torchvision/transforms/functional_pil.py#L73-L79

However, I've tested with pillow >= 8. I don't know about previous version though.

Comment (vfdev-5) [reply torchvision/transforms/functional.py:845]: Well, my request here is to change the sentence below which says that PIL Image model "L" is not supported.

**Comment by vfdev-5** [comment]
@VinhLoiIT @datumbox what about `AutoAugment`, does it support gray images now with this PR ? If yes, let's update the doc as well

**Comment by VinhLoiIT** [comment]
@vfdev-5 I'm writing tests for that now. 

Currently, the tests use `fill=(128,128,128)`:
https://github.com/pytorch/vision/blob/4ef8e6bc73368c21654b9cae95fcfbab04777bf8/test/test_transforms.py#L1494-L1503
which conflicts with my proposed changes as following:
```python
@pytest.mark.parametrize('num_ops', [1, 2, 3])
@pytest.mark.parametrize('magnitude', [7, 9, 11])
@pytest.mark.parametrize('fill', [None, 85, (128, 128, 128)])
@pytest.mark.parametrize('grayscale', [True, False])
def test_randaugment(num_ops, magnitude, fill, grayscale):
    random.seed(42)
    img = Image.open(GRACE_HOPPER)
    if grayscale:
        img = img.convert('L')
    # conflict when fill=(128,128,128) and grayscale=True
    transform = transforms.RandAugment(num_ops=num_ops, magnitude=magnitude, fill=fill)
    for _ in range(100):
        img = transform(img)
    transform.__repr__()
```

Do you have any suggestion here?

**Comment by vfdev-5** [comment]
@VinhLoiIT how about doing the following:
```python
    if grayscale:
        img = img.convert('L')
        fill = (fill[0], ) if isinstance(fill, tuple) else fill
```

**Comment by VinhLoiIT** [comment]
@vfdev-5 Okay it works now. I'll update the tests soon.

File: torchvision/transforms/functional.py (hunk 6 above)
Comment (vfdev-5) [reply torchvision/transforms/functional.py:845]: @VinhLoiIT please address these unresolved conversations

Comment (VinhLoiIT) [reply torchvision/transforms/functional.py:845]: Yes, I've updated the docstring of the ColorJitter class

Comment (vfdev-5) [reply torchvision/transforms/functional.py:845]: This one is for `adjust_hue`

**Comment by datumbox** [comment]
@VinhLoiIT FYI, the failing tests seems related.

File: test/test_functional_tensor.py [hunk 7]
```diff
@@ -756,34 +766,39 @@ def test_solarize2(device, dtype, config):
 @pytest.mark.parametrize('device', cpu_and_gpu())
 @pytest.mark.parametrize('dtype', (None, torch.float32, torch.float64))
 @pytest.mark.parametrize('config', [{"sharpness_factor": f} for f in [0.2, 0.5, 1.0, 1.5, 2.0]])
-def test_adjust_sharpness(device, dtype, config):
+@pytest.mark.parametrize('channels', [1, 3])
+def test_adjust_sharpness(device, dtype, config, channels):
```
Comment (vfdev-5) [inline test/test_functional_tensor.py:770]: This test seems failing: https://app.circleci.com/pipelines/github/pytorch/vision/10743/workflows/fb105be8-f968-493f-9bf4-e3f22251eec5/jobs/819925/tests

Comment (VinhLoiIT) [reply test/test_functional_tensor.py:770]: Well, when I run this test locally with the command:
```
pytest test/test_functional_tensor.py -vvv -k test_adjust_sharpness
```

It shows:
```
========================================================================================== test session starts ==========================================================================================
platform linux -- Python 3.8.11, pytest-6.2.4, py-1.10.0, pluggy-0.13.1 -- /home/vinhloiit/miniconda3/envs/cv/bin/python
cachedir: .pytest_cache
rootdir: /home/vinhloiit/projects/vision, configfile: pytest.ini
plugins: cov-2.12.1, anyio-2.2.0
collected 28414 items / 28354 deselected / 60 selected                                                                                                                                                  

test/test_functional_tensor.py::test_adjust_sharpness[1-config0-None-cpu] PASSED                                                                                                                  [  1%]
test/test_functional_tensor.py::test_adjust_sharpness[1-config0-None-cuda] PASSED                                                                                                                 [  3%]
test/test_functional_tensor.py::test_adjust_sharpness[1-config0-dtype1-cpu] PASSED                                                                                                                [  5%]
test/test_functional_tensor.py::test_adjust_sharpness[1-config0-dtype1-cuda] PASSED                                                                                                               [  6%]
test/test_functional_tensor.py::test_adjust_sharpness[1-config0-dtype2-cpu] PASSED                                                                                                                [  8%]
test/test_functional_tensor.py::test_adjust_sharpness[1-config0-dtype2-cuda] PASSED                                                                                                               [ 10%]
test/test_functional_tensor.py::test_adjust_sharpness[1-config1-None-cpu] PASSED                                                                                                                  [ 11%]
test/test_functional_tensor.py::test_adjust_sharpness[1-config1-None-cuda] PASSED                                                                                                                 [ 13%]
test/test_functional_tensor.py::test_adjust_sharpness[1-config1-dtype1-cpu] PASSED                                                                                                                [ 15%]
test/test_functional_tensor.py::test_adjust_sharpness[1-config1-dtype1-cuda] PASSED                                                                                                               [ 16%]
test/test_functional_tensor.py::test_adjust_sharpness[1-config1-dtype2-cpu] PASSED                                                                                                                [ 18%]
test/test_functional_tensor.py::test_adjust_sharpness[1-config1-dtype2-cuda] PASSED                                                                                                               [ 20%]
test/test_functional_tensor.py::test_adjust_sharpness[1-config2-None-cpu] PASSED                                                                                                                  [ 21%]
test/test_functional_tensor.py::test_adjust_sharpness[1-config2-None-cuda] PASSED                                                                                                                 [ 23%]
test/test_functional_tensor.py::test_adjust_sharpness[1-config2-dtype1-cpu] PASSED                                                                                                                [ 25%]
test/test_functional_tensor.py::test_adjust_sharpness[1-config2-dtype1-cuda] PASSED                                                                                                               [ 26%]
test/test_functional_tensor.py::test_adjust_sharpness[1-config2-dtype2-cpu] PASSED                                                                                                                [ 28%]
test/test_functional_tensor.py::test_adjust_sharpness[1-config2-dtype2-cuda] PASSED                                                                                                               [ 30%]
test/test_functional_tensor.py::test_adjust_sharpness[1-config3-None-cpu] PASSED                                                                                                                  [ 31%]
test/test_functional_tensor.py::test_adjust_sharpness[1-config3-None-cuda] PASSED                                                                                                                 [ 33%]
test/test_functional_tensor.py::test_adjust_sharpness[1-config3-dtype1-cpu] PASSED                                                                                                                [ 35%]
test/test_functional_tensor.py::test_adjust_sharpness[1-config3-dtype1-cuda] PASSED                                                                                                               [ 36%]
test/test_functional_tensor.py::test_adjust_sharpness[1-config3-dtype2-cpu] PASSED                                                                                                                [ 38%]
test/test_functional_tensor.py::test_adjust_sharpness[1-config3-dtype2-cuda] PASSED                                                                                                               [ 40%]
test/test_functional_tensor.py::test_adjust_sharpness[1-config4-None-cpu] PASSED                                                                                                                  [ 41%]
test/test_functional_tensor.py::test_adjust_sharpness[1-config4-None-cuda] PASSED                                                                                                                 [ 43%]
test/test_functional_tensor.py::test_adjust_sharpness[1-config4-dtype1-cpu] PASSED                                                                                                                [ 45%]
test/test_functional_tensor.py::test_adjust_sharpness[1-config4-dtype1-cuda] PASSED                                                                                                               [ 46%]
test/test_functional_tensor.py::test_adjust_sharpness[1-config4-dtype2-cpu] PASSED                                                                                                                [ 48%]
test/test_functional_tensor.py::test_adjust_sharpness[1-config4-dtype2-cuda] PASSED                                                                                                               [ 50%]
test/test_functional_tensor.py::test_adjust_sharpness[3-config0-None-cpu] PASSED                                                                                                                  [ 51%]
test/test_functional_tensor.py::test_adjust_sharpness[3-config0-None-cuda] PASSED                                                                                                                 [ 53%]
test/test_functional_tensor.py::test_adjust_sharpness[3-config0-dtype1-cpu] PASSED                                                                                                                [ 55%]
test/test_functional_tensor.py::test_adjust_sharpness[3-config0-dtype1-cuda] PASSED                                                                                                               [ 56%]
test/test_functional_tensor.py::test_adjust_sharpness[3-config0-dtype2-cpu] PASSED                                                                                                                [ 58%]
test/test_functional_tensor.py::test_adjust_sharpness[3-config0-dtype2-cuda] PASSED                                                                                                               [ 60%]
test/test_functional_tensor.py::test_adjust_sharpness[3-config1-None-cpu] PASSED                                                                                                                  [ 61%]
test/test_functional_tensor.py::test_adjust_sharpness[3-config1-None-cuda] PASSED                                                                                                                 [ 63%]
test/test_functional_tensor.py::test_adjust_sharpness[3-config1-dtype1-cpu] PASSED                                                                                                                [ 65%]
test/test_functional_tensor.py::test_adjust_sharpness[3-config1-dtype1-cuda] PASSED                                                                                                               [ 66%]
test/test_functional_tensor.py::test_adjust_sharpness[3-config1-dtype2-cpu] PASSED                                                                                                                [ 68%]
test/test_functional_tensor.py::test_adjust_sharpness[3-config1-dtype2-cuda] PASSED                                                                                                               [ 70%]
test/test_functional_tensor.py::test_adjust_sharpness[3-config2-None-cpu] PASSED                                                                                                                  [ 71%]
test/test_functional_tensor.py::test_adjust_sharpness[3-config2-None-cuda] PASSED                                                                                                                 [ 73%]
test/test_functional_tensor.py::test_adjust_sharpness[3-config2-dtype1-cpu] PASSED                                                                                                                [ 75%]
test/test_functional_tensor.py::test_adjust_sharpness[3-config2-dtype1-cuda] PASSED                                                                                                               [ 76%]
test/test_functional_tensor.py::test_adjust_sharpness[3-config2-dtype2-cpu] PASSED                                                                                                                [ 78%]
test/test_functional_tensor.py::test_adjust_sharpness[3-config2-dtype2-cuda] PASSED                                                                                                               [ 80%]
test/test_functional_tensor.py::test_adjust_sharpness[3-config3-None-cpu] PASSED                                                                                                                  [ 81%]
test/test_functional_tensor.py::test_adjust_sharpness[3-config3-None-cuda] PASSED                                                                                                                 [ 83%]
test/test_functional_tensor.py::test_adjust_sharpness[3-config3-dtype1-cpu] PASSED                                                                                                                [ 85%]
test/test_functional_tensor.py::test_adjust_sharpness[3-config3-dtype1-cuda] PASSED                                                                                                               [ 86%]
test/test_functional_tensor.py::test_adjust_sharpness[3-config3-dtype2-cpu] PASSED                                                                                                                [ 88%]
test/test_functional_tensor.py::test_adjust_sharpness[3-config3-dtype2-cuda] PASSED                                                                                                               [ 90%]
test/test_functional_tensor.py::test_adjust_sharpness[3-config4-None-cpu] PASSED                                                                                                                  [ 91%]
test/test_functional_tensor.py::test_adjust_sharpness[3-config4-None-cuda] PASSED                                                                                                                 [ 93%]
test/test_functional_tensor.py::test_adjust_sharpness[3-config4-dtype1-cpu] PASSED                                                                                                                [ 95%]
test/test_functional_tensor.py::test_adjust_sharpness[3-config4-dtype1-cuda] PASSED                                                                                                               [ 96%]
test/test_functional_tensor.py::test_adjust_sharpness[3-config4-dtype2-cpu] PASSED                                                                                                                [ 98%]
test/test_functional_tensor.py::test_adjust_sharpness[3-config4-dtype2-cuda] PASSED                                                                                                               [100%]
```

File: test/test_transforms.py [hunk 8]
```diff
@@ -1482,9 +1482,13 @@ def test_five_crop(single_dim):
...
 @pytest.mark.parametrize('policy', transforms.AutoAugmentPolicy)
 @pytest.mark.parametrize('fill', [None, 85, (128, 128, 128)])
-def test_autoaugment(policy, fill):
+@pytest.mark.parametrize('grayscale', [True, False])
+def test_autoaugment(policy, fill, grayscale):
     random.seed(42)
     img = Image.open(GRACE_HOPPER)
+    if grayscale:
+        img = img.convert('L')
+        fill = (fill[0], ) if isinstance(fill, tuple) else fill
```
Comment (datumbox) [inline test/test_transforms.py:1491]: Should we move this to a private util method to avoid code duplication?

**Review by datumbox — COMMENTED** [review]
Thanks for the PR @VinhLoiIT.

The test failures look like a tolerance problem and might be resolved by adjusting it.

File: test/test_functional_tensor.py (hunk 7 above)
Comment (vfdev-5) [reply test/test_functional_tensor.py:770]: I could not repro neither locally from my side using the same pytorch. According to the failure report, the difference is
```
AssertionError: Tensor-likes are not equal!

Mismatched elements: 56 / 288 (19.4%)
Greatest absolute difference: 5.960464477539063e-08 at index (0, 5, 4)
Greatest relative difference: 1.47366798466619e-07 at index (0, 2, 4)
```

After debugging on Circle CI, max abs error is about 5.9605e-08 for all failing cases. Looks like a flakyness on this particular infrastructure

I propose to do the following fix : 
```diff
# vision/test/common_utils.py
- assert_equal = functools.partial(torch.testing.assert_close, rtol=0, atol=0)
+ assert_equal = functools.partial(torch.testing.assert_close, rtol=0, atol=1e-6)
```

File: test/test_transforms.py (hunk 8 above)
Comment (VinhLoiIT) [reply test/test_transforms.py:1491]: I suggest defining an util function at the top of the file as follow:

```python
def get_grayscale_test_image(img, fill=None):
    img = img.convert('L')
    fill = (fill[0], ) if isinstance(fill, tuple) else fill
    return img, fill
```
And, for example, update those test cases as below:
```python
@pytest.mark.parametrize('num_ops', [1, 2, 3])
@pytest.mark.parametrize('magnitude', [7, 9, 11])
@pytest.mark.parametrize('fill', [None, 85, (128, 128, 128)])
@pytest.mark.parametrize('grayscale', [True, False])
def test_randaugment(num_ops, magnitude, fill, grayscale):
    random.seed(42)
    img = Image.open(GRACE_HOPPER)
    img, fill = get_grayscale_test_image(img, fill)  # change here
    transform = transforms.RandAugment(num_ops=num_ops, magnitude=magnitude, fill=fill)
    for _ in range(100):
        img = transform(img)
    transform.__repr__()
```

What do you think?

Comment (datumbox) [reply test/test_transforms.py:1491]: Sounds good to me. Just make it private by prepending `_` to it. :)

**Review by datumbox — APPROVED** [review]
LGTM from my side, I'll leave it to @vfdev-5 for the final approval and merge.

Thanks for the contribution @VinhLoiIT!

**Review by vfdev-5 — APPROVED** [review]
Looks good to me as well!
@VinhLoiIT thanks for the PR!
